sdk/eventhub/event-hubs/src/partitionPump.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as log from "./log";
+import { EventProcessorOptions, PartitionProcessor } from "./eventProcessor";
+import { PartitionContext } from "./partitionContext";
+import { EventHubClient } from "./eventHubClient";
+import { EventPosition } from "./eventPosition";
+import { EventHubConsumer } from "./receiver";
+
+export class PartitionPump {
+  private _partitionContext: PartitionContext;
+  private _eventHubClient: EventHubClient;
+  private _partitionProcessor: PartitionProcessor;
+  private _processorOptions: EventProcessorOptions;
+  private _receiver: EventHubConsumer | undefined;
+  private _isReceiving: boolean = false;
+
+  constructor(
+    eventHubClient: EventHubClient,
+    partitionContext: PartitionContext,
+    partitionProcessor: PartitionProcessor,
+    options?: EventProcessorOptions
+  ) {
+    if (!options) options = {};
+    this._eventHubClient = eventHubClient;
+    this._partitionContext = partitionContext;
+    this._partitionProcessor = partitionProcessor;
+    this._processorOptions = options;
+  }
+
+  async start(partitionId: string): Promise<void> {
+    await this._partitionProcessor.initialize!();
+    await this._receiveEvents(partitionId);
+    log.partitionPump("Successfully started the receiver.");
+  }
+
+  private async _receiveEvents(partitionId: string): Promise<void> {
+    this._isReceiving = true;
+    this._receiver = await this._eventHubClient.createConsumer(
+      this._partitionContext.consumerGroupName,
+      partitionId,
+      this._processorOptions.initialEventPosition || EventPosition.earliest()
+    );
+    try {
+      while (this._isReceiving) {
+        const receivedEvents = await this._receiver.receiveBatch(
+          this._processorOptions.maxBatchSize || 1,
+          this._processorOptions.maxWaitTime
+        );
+        await this._partitionProcessor.processEvents(receivedEvents);
+      }
+    } catch (err) {
+      this._isReceiving = false;
+      this._receiver.close();
+      await this._partitionProcessor.processError(err);
+      log.partitionPump("An error occurred while receiving events.", err);
+    }
+  }
+
+  async stop(): Promise<void> {
+    this._isReceiving = false;
+    try {
+      if (this._receiver) {
+        this._receiver.close();
```

Comment:
Is your intent to kick off the `close` and then immediately go to the customer's `close` method?

If so, we should also add a `.catch` to `close()` so we swallow any errors encountered with that. Otherwise we might get an UnhandledPromiseRejectionWarning.

One thing I think we'll have to spend more time on is guaranteeing we close the underlying AMQP link. I think right now if we were to quickly call `start` and then `stop`, we might try closing the AMQP link while it's being initialized and up with a link we can't close.

Reply:
Yeah..I'll test that scenario.

Reply:
If each `partitionPump` had an internal `abortController`, it could pass the `abortSignal` to the `receiveBatch` call, and call `abort` when `stop` is called. I think that would ensure the link is closed. This would mean we also need to avoid calling the `processError` handler if we stopped receiving messages so the `AbortError` doesn't get forwarded.

Reply:
>One thing I think we'll have to spend more time on is guaranteeing we close the underlying AMQP link. I think right now if we were to quickly call start and then stop, we might try closing the AMQP link while it's being initialized and end up with a link we can't close.

This is the generic issue, I think we should handle this scenario inside close().

For now I'll pass the abortSignal to the receiveBatch call, and call abort when stop is called.